### ISSUE TITLE
How to retrieve all items from a mapping in efficient manner

### ISSUE THREAD

**Issue by swswsw** [Encode-Club-Solidity-Foundations/Lesson-02#1]
for example:

```
    struct Item {
        uint id;
        string nameOfItem;
        string typeofItem;
        uint value;
    }
    
   struct User {
        address userAddress;
        string email;
    }

    mapping (address => Item[]) public userItems;

```

It is easy to get Item based given user address.  What if we want to get all user address?  Or if we want to get all the items?

**Comment by MatheusDaros** [comment]
Hello @swswsw 
Thanks for the question.
Unfortunately it is impossible to directly return all the elements in a mapping or even in a array using solidity.
A possible workaround would be keeping an array to track all the addresses that have been populated inside the mapping, and then iterating through the array to get all elements.

**Comment by luksgrin** [comment]
It could be possible to implement a library called "extended mapping", consisting of a mapping with extended methods such as `python`'s `.items()`, `.keys()`, `.values()`...

**Comment by MatheusDaros** [comment]
Yes, it would. The problem is that those constructions spend huge amounts of space to map it all out. It is advisable to, whenever possible, do all of these operations offchain, and include onchain only the inevitable.
